design/FY2018/NFP-Table-Refactor.md

```diff
@@ -0,0 +1,819 @@
+
+#### Linear Interpolation
+
+For linear interpolation in 1 dimension, given known values of $f$ at $x_0$ and $x_1$, and a point $x$ between $x_0$ and $x_1$, the value at $x$ is estimated by:
+
+$$f\left(x\right) = \left(1-\mu\right) \cdot f\left(x_0\right) + \mu \cdot f\left(x_1\right)$$
+
+where $\mu$ is the location of $x$ expressed as a fraction of the distance between $x_0$ and $x_1$:
+
+$$\mu = \left(x-x_0\right) / \left(x_1 - x_0\right)$$
+
+
+#### Cubic Spline Interpolation
+
+The general formula for a 1-dimensional piecewise cubic spline (for $x$ between known values $f\left(x_0\right)$ and $f\left(x_1\right)$) is:
+
+$$\begin{array}{rll}
+f\left(x\right) &= \left(2\mu^3 - 3\mu^2 + 1\right) \cdot f\left(x_0\right)
+&+ \left(-2\mu^3 + 3\mu^2\right) \cdot f\left(x_1\right) \\
+&+ \left(\mu^3 - 2\mu^2 + \mu\right) \cdot f^\prime \left(x_0\right)
+&+ \left(\mu^3 - \mu^2\right) \cdot f^\prime \left(x_1\right)
+\end{array}$$
+
+The Catmull-Rom cubic spline interpolation defines the derivatives $f^\prime \left(x_0\right)$ and $f^\prime\left(x_1\right)$ as the slope between the previous and following grid points on the axis:
+
+$$\begin{array}{rl}
+f^\prime(x_0) &= ( f(x_1) - f(x_{-1}) ) / ( x_1 - x_{-1}) \\
+f^\prime(x_1) &= ( f(x_2) - f(x_0) ) / ( x_2 - x_0)
+\end{array}$$
+
+When the hypercube is at the edge of the grid, Catmull-Rom simply extends the slope of the final segment for defining the slope terms: i.e., if there is no $x_{-1}$, we substitute $x_0$ into the $f^\prime(x_0)$ formula.
+
+If a mix of interpolation methods are specified among the included dimensions, the library will perform that mix as requested.
+
+If the lookup point is beyond the grid edge on any axis, the library will perform the requested extrapolation method (linear or constant) on that dimension while proceeding with interpolation along any in-bounds dimension. If a "do-not-extrapolate-beyond" boundary is specified, the library will perform a constant extrapolation from that boundary outward (in that dimension), and it will return the resulting numerical answer along with a warning that the request was outside the boundaries.
+
+
+## Source Code Design
+
+The main component of the source code design is the addition of a new tabular data interpolation library (currently called "Btwxt" [betwixt]). The purpose of this library is to isolate the interpolation computation and data structures from the rest of EnergyPlus in a fashion that is easily unit tested and can be easily ported to other C++ programs.
+
+The other component is the integration of Btwxt in EnergyPlus. This will involve a modest refactoring of how EnergyPlus handles table data structures.
+
+### Btwxt
```

Comment:
@nealkruis Is this an existing third-party library or a new library developed for this work?

Reply:
This is new library developed by Big Ladder, I'll push a version up into the repo today.